### PR TITLE
chore: refactor to track spans with location

### PR DESCRIPTION
**Pull request by kobyhallx**
# Description

## Problem\*

Resolves 

Locations of language constructs lack full location information #4167

## Summary\*

LSP operations like goto, refactoring (ie. name change) or finding references need to understand where something was defined or where it is referenced. To do that (understand references or definitions) lsp needs to see not only Spans which define type or function but also context in which those spans exist. That context can be File if a tool operates within file system, url if source files are downloaded from url or buffers if they are supplied through api.

Spans gained context from which they are being extracted - SrcId, which one possibility of is FileId in case context is file system.

## Additional Context

This PR refactors how Spans are collected by placing them in context of a src_id. LIbrary of choice - `chumsky` already provides ability to place Spans in context by defining [`type Context`](https://github.com/zesterer/chumsky/blob/b6afd8c738799060be8b43e2e618247d360a45eb/src/span.rs#L23) on Spans trait with documentation explaining:
```
    /// Extra context used in a span.
    ///
    /// This is usually some way to uniquely identity the source file that a span originated in such as the file's
    /// path, URL, etc.
```
by utilising this construct we're able to attach this information from the very beginning of the collection process and carry it through to places where they are needed.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_errors/src/position.rs [hunk 1]
```diff
@@ -6,7 +6,37 @@ use std::{
...
+    fn add(self, rhs: u32) -> Position {
+        Position(self.0 + rhs, self.1)
+    }
+}
+
+impl From<Position> for u32 {
+    fn from(val: Position) -> Self {
+        val.0
+    }
+}
```
Comment (jfecher) [inline compiler/noirc_errors/src/position.rs:33]: I think this impl is unclear in a struct with two Fields, even if one of those Fields is directly a `u32`. I'd prefer an explicit conversion like `position.index()` or position_index or similar.

File: compiler/noirc_errors/src/position.rs [hunk 2]
```diff
@@ -51,27 +81,49 @@ impl<T> std::borrow::Borrow<T> for Spanned<T> {
...
+    fn from(value: usize) -> Self {
+        Self(value)
+    }
+}
+
+impl From<SrcId> for usize {
+    fn from(val: SrcId) -> Self {
+        val.0
+    }
+}
```
Comment (jfecher) [inline compiler/noirc_errors/src/position.rs:99]: What does a `SrcId` represent? There are no comments.

Similarly to my other comments, lets remove the `From` impls here as well. I think they just obfuscate things here. Using methods we can also control the visibility more tightly, e.g. making constructing SrcIds private.

File: compiler/noirc_errors/src/position.rs [hunk 3]
```diff
@@ -51,27 +81,49 @@ impl<T> std::borrow::Borrow<T> for Spanned<T> {
...
+
+    pub fn inclusive_within(start: u32, end: u32, src_id: impl Into<usize>) -> Span {
+        Span(ByteSpan::from(start..end + 1), src_id.into().into())
     }
 
-    pub fn single_char(start: u32) -> Span {
-        Span::inclusive(start, start)
+    pub fn single_char(start: u32, src_id: impl Into<usize>) -> Span {
+        let start_position = Position::new(start, src_id.into().into());
+        Span::inclusive(start_position, start_position)
```
Comment (jfecher) [inline compiler/noirc_errors/src/position.rs:117]:
```suggestion
    pub fn single_char(start: u32, src_id: SrcId) -> Span {
        let start_position = Position::new(start, src_id);
        Span::inclusive(start_position, start_position)
```
`.into().into()` is definitely too far for me. If we need a `SrcId`, lets have the function take one directly

File: compiler/noirc_errors/src/position.rs [hunk 4]
```diff
@@ -51,27 +81,49 @@ impl<T> std::borrow::Borrow<T> for Spanned<T> {
...
 
 impl Span {
-    pub fn inclusive(start: u32, end: u32) -> Span {
-        Span(ByteSpan::from(start..end + 1))
+    pub fn inclusive(start: Position, end: Position) -> Span {
+        Span(ByteSpan::from(start.0..end.0 + 1), start.1)
+    }
+
+    pub fn inclusive_within(start: u32, end: u32, src_id: impl Into<usize>) -> Span {
+        Span(ByteSpan::from(start..end + 1), src_id.into().into())
```
Comment (jfecher) [inline compiler/noirc_errors/src/position.rs:112]: Same here

File: compiler/noirc_errors/src/position.rs [hunk 5]
```diff
@@ -51,27 +81,49 @@ impl<T> std::borrow::Borrow<T> for Spanned<T> {
...
+    fn from(val: SrcId) -> Self {
+        val.0
+    }
+}
+
 #[derive(
     PartialEq, PartialOrd, Eq, Ord, Hash, Debug, Copy, Clone, Default, Deserialize, Serialize,
 )]
-pub struct Span(ByteSpan);
+pub struct Span(ByteSpan, SrcId);
```
Comment (jfecher) [inline compiler/noirc_errors/src/position.rs:104]: If a `SrcId` is meant to represent a File, do we need `Location` at all anymore?

I've yet to see this whole PR yet so I don't know how a SrcId would correspond to a FileId yet, e.g. if they're convertible or not.

File: compiler/noirc_errors/src/position.rs [hunk 6]
```diff
@@ -105,20 +161,22 @@ impl From<Span> for Range<usize> {
 
 impl From<Range<u32>> for Span {
     fn from(Range { start, end }: Range<u32>) -> Self {
-        Self(ByteSpan::new(start, end))
+        Self(ByteSpan::new(start, end), SrcId::default())
```
Comment (jfecher) [inline compiler/noirc_errors/src/position.rs:164]: The `SrcId::default()` here is definitely suspicious, and likely incorrect. I think we can't have a `From<Range<u32>>` impl anymore if we need a SrcId as well.

File: compiler/noirc_errors/src/position.rs [hunk 7]
```diff
@@ -51,27 +81,49 @@ impl<T> std::borrow::Borrow<T> for Spanned<T> {
     }
 }
 
+#[derive(
+    Copy, Clone, Serialize, Eq, PartialEq, Ord, PartialOrd, Debug, Deserialize, Hash, Default,
```
Comment (jfecher) [inline compiler/noirc_errors/src/position.rs:85]: `Default` seems like it could possibly introduce bugs to me. I'm unsure if `Default` would always point to src/main.rs or a random file that happens to be first but I don't think we should need a constructor to return either of those two cases.

File: compiler/noirc_frontend/src/hir/def_map/mod.rs [hunk 8]
```diff
@@ -296,7 +296,7 @@ pub struct Contract {
 /// Given a FileId, fetch the File, from the FileManager and parse it's content
 pub fn parse_file(fm: &FileManager, file_id: FileId) -> (ParsedModule, Vec<ParserError>) {
     let file_source = fm.fetch_file(file_id).expect("File does not exist");
-    parse_program(file_source)
+    parse_program(file_id.as_usize().into(), file_source)
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/def_map/mod.rs:299]: Why don't we use `FileId`s directly instead of source ids?

File: compiler/noirc_frontend/src/hir/type_check/mod.rs [hunk 9]
```diff
@@ -479,7 +479,7 @@ mod test {
         expected_number_errors: usize,
         func_namespace: Vec<String>,
     ) {
-        let (program, errors) = parse_program(src);
+        let (program, errors) = parse_program(Default::default(), src);
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/type_check/mod.rs:482]: If `SrcId: Default` is only used for testing we could have a test-only function instead:
```rs
impl SrcId {
    #[cfg(test)]
    fn no_source() -> Self { ... }
}
```

File: compiler/noirc_frontend/src/lexer/lexer.rs [hunk 10]
```diff
@@ -17,15 +17,23 @@ pub struct Lexer<'a> {
...
+    src_id: SrcId,
 }
 
 pub type SpannedTokenResult = Result<SpannedToken, LexerErrorKind>;
 
 impl<'a> Lexer<'a> {
+    /// Given a source buffer of noir code, return all the tokens in the buffer
+    /// in order, along with any lexing errors that occurred.
+    pub fn lex_source(source: &'a str) -> (Tokens, Vec<LexerErrorKind>) {
+        let src_id = SrcId::default();
```
Comment (jfecher) [inline compiler/noirc_frontend/src/lexer/lexer.rs:29]: I'm guessing this should be a `#[cfg(test)]` function as well.

Also, since we don't pass a `SrcId` to this function I think `lex_no_source` or `lex_raw_string` would be better names.

File: compiler/noirc_frontend/src/lexer/lexer.rs [hunk 11]
```diff
@@ -37,13 +45,18 @@ impl<'a> Lexer<'a> {
         (Tokens(tokens), errors)
     }
 
-    pub fn new(source: &'a str) -> Self {
+    pub fn single_source(source: &'a str) -> Self {
+        Self::new(SrcId::default(), source)
+    }
```
Comment (jfecher) [inline compiler/noirc_frontend/src/lexer/lexer.rs:50]:
```suggestion
    #[cfg(test)]
    pub fn from_raw_string(source: &'a str) -> Self {
        Self::new(SrcId::default(), source)
    }
```
I wasn't sure on the `single_source` name either

**Comment by kobyhallx** [comment]
Thanks for v detailed review though I specifically marked it draft to seek agreement on general direction rather than those minor points (though equally appreciated).


> - I think `From`/`Into` are used a bit too much in this PR, hurting readability

this mostly for keeping api surface more convenient. Don't have a strong preference either way. Will look for a more balanced approach.


> - I don't understand why we are passing around a new `SrcId` tag instead of the existing `FileId`. It looks like we're converting from `FileId` to `SrcId` when constructing them, so we do have them available. Am I missing something?

I can imagine a context where this is not within file system context and srcid seems more appropriate when we would not drag the file manager crate. Though this is perhaps just naming and location, where we probably want srcid/fileid closer to parser than file manager. File manager should then issue parser's srcid's.

> - If `Span`s have a `SrcId`, we can probably get rid of `Location`, right?

yes,  I  was not sure why Location got introduced this way in the first place, and that's where I needed feedback most, from this comment I will continue to get rid of it.

> - We should rename Span to Location since it carries more than just span information now.

I think Span is named correctly and follows the base library nomenclature - as we should argue about Spans within context (of url, file, etc.). Sometimes we would argue about a source w/o context...

**Comment by jfecher** [comment]
> I think Span is named correctly and follows the base library nomenclature - as we should argue about Spans within context (of url, file, etc.). Sometimes we would argue about a source w/o context...

Can you elaborate on this point?

File: compiler/noirc_errors/src/position.rs (hunk 5 above)
Comment (kobyhallx) [reply compiler/noirc_errors/src/position.rs]: Location removed

File: compiler/noirc_errors/src/position.rs (hunk 4 above)
Comment (kobyhallx) [reply compiler/noirc_errors/src/position.rs]: replaced with srcid

File: compiler/noirc_errors/src/position.rs (hunk 3 above)
Comment (kobyhallx) [reply compiler/noirc_errors/src/position.rs:120]: applied

File: compiler/noirc_errors/src/position.rs (hunk 6 above)
Comment (kobyhallx) [reply compiler/noirc_errors/src/position.rs]: `From<Range<u32>>` moved to fmt module

File: compiler/noirc_errors/src/position.rs (hunk 1 above)
Comment (kobyhallx) [reply compiler/noirc_errors/src/position.rs:33]: 9e63896: chore: clearer Position naming

File: compiler/noirc_frontend/src/lexer/lexer.rs (hunk 10 above)
Comment (kobyhallx) [reply compiler/noirc_frontend/src/lexer/lexer.rs:29]: removed

File: compiler/fm/src/file_map.rs [hunk 12]
```diff
@@ -53,6 +53,12 @@ impl FileId {
     }
 }
 
+impl From<FileId> for usize {
+    fn from(file_id: FileId) -> Self {
+        file_id.0
+    }
+}
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply compiler/fm/src/file_map.rs]: moved to use SrcId instead

File: compiler/noirc_errors/src/position.rs (hunk 2 above)
Comment (kobyhallx) [reply compiler/noirc_errors/src/position.rs]: documented, impl From<usize> dropped

File: compiler/noirc_errors/src/position.rs (hunk 7 above)
Comment (kobyhallx) [reply compiler/noirc_errors/src/position.rs:85]: I'm fine removing defaults but this leads to other refactorings, eg
[unwrap_or_default](https://github.com/noir-lang/noir/blob/1d177d2ac8eb1400718a35203fd04d8134e352c1/compiler/noirc_frontend/src/hir/def_collector/errors.rs#L251)
...where at the end of the day we need to provide some kind of default or be more strictier if not definiing source of something is ok to do.

File: compiler/noirc_frontend/src/hir/def_map/mod.rs (hunk 8 above)
Comment (kobyhallx) [reply compiler/noirc_frontend/src/hir/def_map/mod.rs]: reworked to use SrcId across instead

File: compiler/noirc_frontend/src/hir/type_check/mod.rs (hunk 9 above)
Comment (kobyhallx) [reply compiler/noirc_frontend/src/hir/type_check/mod.rs]: provided no_source function

File: compiler/noirc_frontend/src/lexer/lexer.rs (hunk 11 above)
Comment (kobyhallx) [reply compiler/noirc_frontend/src/lexer/lexer.rs]: renamed to `from_string`

**Comment by kobyhallx** [comment]
> > I think Span is named correctly and follows the base library nomenclature - as we should argue about Spans within context (of url, file, etc.). Sometimes we would argue about a source w/o context...
> 
> Can you elaborate on this point?

As we read library code they named this `Span` and then provided option to represent this Span within context, which is correct. We can then argue about `Span`s themselves within context. Naming it Location seems to lead to point in space rather than something that represents a Span.
Either way is fine for me but `Span` within context sound better to me.

**Comment by TomAFrench** [comment]
Agree on `Span` not being correctly named. I'm also not sure on the need to squash `Location` and `Span` together as we have, this has led to some oddities such as being able to merge two `Spans` which live in separate files.

**Comment by kobyhallx** [comment]
> I'm also not sure on the need to squash `Location` and `Span` together as we have, this has led to some oddities such as being able to merge two `Spans` which live in separate files.

You can argue both ways about that, to simplify our case, added checks here 
81dd0af: chore: prevent cross file spans

**Comment by kevaundray** [comment]
This PR seems to be doing multiple changes, are all of these necesarry to solve the original issue?

File: compiler/noirc_driver/src/lib.rs [hunk 13]
```diff
@@ -275,13 +275,13 @@ pub fn compile_contract(
...
         let err = CustomDiagnostic::from_message("Packages are limited to a single contract")
-            .in_file(FileId::default());
+            .in_file(SrcId::default());
         return Err(vec![err]);
     } else if contracts.is_empty() {
         let err = CustomDiagnostic::from_message(
             "cannot compile crate into a contract as it does not contain any contracts",
         )
-        .in_file(FileId::default());
+        .in_file(SrcId::default());
```
Comment (kevaundray) [inline compiler/noirc_driver/src/lib.rs:284]: This rename looks like it can be in a separate PR

File: compiler/noirc_errors/src/debug_info.rs [hunk 14]
```diff
@@ -27,7 +27,7 @@ pub struct DebugInfo {
     /// Serde does not support mapping keys being enums for json, so we indicate
     /// that they should be serialized to/from strings.
     #[serde_as(as = "BTreeMap<DisplayFromStr, _>")]
-    pub locations: BTreeMap<OpcodeLocation, Vec<Location>>,
+    pub locations: BTreeMap<OpcodeLocation, Vec<Span>>,
```
Comment (kevaundray) [inline compiler/noirc_errors/src/debug_info.rs:30]: Before merging I think a rationale of this change would be useful. A span and a location are different things

**Review by kevaundray — CHANGES_REQUESTED** [review]
Read through the PR and a part from the lack of rationale, its not really clear what problem is being solved and why changes/generalizations are being made.

**Comment by kobyhallx** [comment]
> This PR seems to be doing multiple changes, are all of these necesarry to solve the original issue?

Not all changes are required, the change that is most desired is to provide information about Span context - SrcId/FileId, rest  of the changes are based on initial feedback or intent to simplify.

**Comment by kobyhallx** [comment]
> ...its not really clear what problem is being solved and why changes/generalizations are being made.
The problem is stated in referenced issue - while working on lsp features, in some cases information about initial location is not carried through from the parser, where it is carried in form of span, in some cases it lacks information from which source context (ie. fileid) it originates and forces changes to thread fileid to functions. Having FileId/SrcId carried with Span it makes it simpler to fill gaps.

File: compiler/noirc_driver/src/lib.rs (hunk 13 above)
Comment (kobyhallx) [reply compiler/noirc_driver/src/lib.rs:284]: We don't really have to change this name either - it just seemed more logical to name it SrcId that is more universal rather than FileId which is fs specific, though data stays the same.

File: compiler/noirc_errors/src/debug_info.rs (hunk 14 above)
Comment (kobyhallx) [reply compiler/noirc_errors/src/debug_info.rs:30]: They do carry same data: 
Location is Span (offsets) + File/SrcId, 
Span is offsets + context (fileid/srcid),

Not attached one way or the other - it follows however concept defined by library.

File: compiler/noirc_errors/src/position.rs (hunk 7 above)
Comment (jfecher) [reply compiler/noirc_errors/src/position.rs:88]: Lets just table this then for now I suppose. This PR is already quite large so we can leave any further refactorings for later.

File: compiler/noirc_errors/src/debug_info.rs (hunk 14 above)
Comment (jfecher) [reply compiler/noirc_errors/src/debug_info.rs:30]: I think @kevaundray here was likely also confused by the fact a Span is usually just a range, where a location also implies a source file or similar. Chumsky's use of a Span with a context was likely just trying to shoehorn a name into a generic type that could have been either a Span or a Location depending on the context, so they had to choose one.

**Comment by jfecher** [comment]
> As we read library code they named this Span and then provided option to represent this Span within context, which is correct. We can then argue about Spans themselves within context. Naming it Location seems to lead to point in space rather than something that represents a Span.
Either way is fine for me but Span within context sound better to me.

I'm not quite sure I understand the argument here. I'll echo @TomAFrench and vote that I still prefer Location for a "byte range + file/source id."

File: .vscode/tasks.json [hunk 15]
```diff
@@ -0,0 +1,34 @@
+{
+	"version": "2.0.0",
+	"tasks": [
+		{
```
Comment (jfecher) [inline .vscode/tasks.json:4]: Was this file accidentally committed? It seems unrelated

**Review by jfecher — COMMENTED** [review]
I'm fine with most of these changes - although I think also renaming `SrcId` back to `FileId` could help reduce the line count on this PR and hopefully make it more palatable to @TomAFrench  and @kevaundray. We could just think of it as an ID for a potentially virtual file.

**Comment by TomAFrench** [comment]
Personally, I feel like this PR adds a lot of churn:
- a lot of variables/fields now being misnamed as `location` where they're now `Span`s
- `Span` has been squashed into `Location` which then requires us to create a new `ContextLessSpan` which hints that we should maintain the old `Span`.

But the issue which it aims to address seems to boil down to "We're using `Span` somewhere where it may be more convenient to have a `Location`. Wouldn't it be easier to just attach `Location`s instead of these quite wide-ranging changes?

This would make it easier to see where the changes are being made / need to be made.

**Comment by jfecher** [comment]
> But the issue which it aims to address seems to boil down to "We're using Span somewhere where it may be more convenient to have a Location". Wouldn't it be easier to just attach Locations instead of these quite wide-ranging changes?

That was the original intent of the changes IIRC. It seemed at the time like we wanted to use Locations by default everywhere so I think that is why the rename took place. I didn't see `ContextLessSpan` in this PR yet but I agree that's a good reason to undo the name change and go back to Span, Location, and FileId. Just with `Location` in more places than it was previously.

**Comment by kobyhallx** [comment]
>  I agree that's a good reason to undo the name change and go back to Span, Location, and FileId. Just with `Location` in more places than it was previously.

This refactor started this way but it was causing much wider changes than what is here now.

I think this is still the cheapest and most flexible option which follows what the intent of lib authors was.

I can rework the naming, ie Span -> Span and rework api to be more intuitive as those seem to be most contentious, but would encourage to stay with all Spans/Locations carrying fileid data from parsing.
